connector/solarwindsentityconnector/internal/attribute_mapper.go

```diff
@@ -0,0 +1,202 @@
+		}
+		entities = append(entities, newEntity)
+	}
+
+	if isSubset(entity.IDs, attrs.Destination) {
+		newEntity, err := createEntity(entity, attrs.Common, attrs.Destination)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create entity for type %s: %w", entityType, err)
+		}
+		entities = append(entities, newEntity)
+	}
+
+	if isSuperSet(entity.IDs, attrs.Common) {
+		newEntity, err := createEntity(entity, attrs.Common)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create entity for type %s: %w", entityType, err)
+		}
+		entities = append(entities, newEntity)
+	}
+
+	if len(entities) == 0 {
+		return nil, fmt.Errorf("no entityConfigs found for entity type %s with attributes %v", entityType, attrs)
+	}
+
+	return entities, nil
+}
+
+func (e *AttributeMapper) getRelationship(relationship *config.RelationshipEvent, attrs Attributes) (*Relationship, error) {
+	source, ok := e.entityConfigs[relationship.Source]
+	if !ok {
+		return nil, fmt.Errorf("unexpected source entity type")
+	}
+
+	dest, ok := e.entityConfigs[relationship.Destination]
+	if !ok {
+		return nil, fmt.Errorf("unexpected destination entity type")
+	}
+
+	if dest.Type == source.Type {
+		if len(attrs.Source) == 0 && len(attrs.Destination) == 0 {
+			return nil, fmt.Errorf("source and destination attributes are empty for same type relationship %s", relationship.Type)
+		}
+	}
+
+	sourceIds, err := getRequiredAttributes(source.IDs, attrs.Common, attrs.Source)
+	if err != nil {
+		return nil, errors.Join(fmt.Errorf("failed to create source entity %s", source.Type), err)
+
+	}
+
+	destIds, err := getRequiredAttributes(dest.IDs, attrs.Common, attrs.Destination)
+	if err != nil {
+		return nil, errors.Join(fmt.Errorf("failed to create destination entity %s", dest.Type), err)
+	}
+
+	r := Relationship{
+		Type: relationship.Type,
+		Source: RelationshipEntity{
+			Type: source.Type,
+			IDs:  sourceIds,
+		},
+		Destination: RelationshipEntity{
+			Type: dest.Type,
+			IDs:  destIds,
+		},
+		Attributes: getOptionalAttributes(relationship.Attributes, attrs.Common),
+	}
+
+	return &r, nil
+}
+
+func createEntity(entity config.Entity, attrs ...map[string]pcommon.Value) (Entity, error) {
+	ids, err := getRequiredAttributes(entity.IDs, attrs...)
+	if err != nil {
+		return Entity{}, fmt.Errorf("failed to get ID attributes for entity %s: %w", entity.Type, err)
+	}
+
+	ea := getOptionalAttributes(entity.Attributes, attrs...)
+
+	return Entity{
+		IDs:        ids,
+		Attributes: ea,
+		Type:       entity.Type,
+	}, nil
+}
+
+func isSubset(superset []string, subset map[string]pcommon.Value) bool {
+	if subset == nil || len(subset) == 0 {
+		return false
+	}
+
+	for key, _ := range subset {
+		found := false
+		for _, entityId := range superset {
+			if entityId == key {
+				found = true
+				break
+			}
+		}
+		if found == false {
+			return false
+		}
+	}
+
+	return true
+}
+
+func isSuperSet(subset []string, superset map[string]pcommon.Value) bool {
```

Comment:
The name of function and the parameters. I find quite confusing.

```suggestion
func containsAllKeys(keys []string, m map[string]pcommon.Value) bool {
```

Reply:
Renamed to ContainsAll